src/views/core.clj

```diff
@@ -2,138 +2,207 @@
+          (on-unauthorized-subscription @view-system view-sig subscriber-key context)
+          nil)))
+    (throw (new Exception (str "Subscription for non-existant view: " view-id)))))
+
+(defn- remove-from-subscribers
+  [view-system view-sig subscriber-key]
+  (-> view-system
+      (update-in [:subscribers view-sig] disj subscriber-key)
+      ; remove view-sig entry if no subscribers. helps prevent the subscribers
+      ; map from e.g. endlessly filling up with all sorts of different
+      ; view-sigs with crazy amounts of only-slightly-varying parameters
+      (update :subscribers
+              (fn [subscribers]
+                (if (empty? (get subscribers view-sig))
+                  (dissoc subscribers view-sig)
+                  subscribers)))))
+
+(defn- remove-from-subscribed
   [view-system view-sig subscriber-key]
-  (update-in view-system [:subscribers view-sig] disj subscriber-key))
+  (-> view-system
+      (update-in [:subscribed subscriber-key] disj view-sig)
+      ; remove subscriber-key entry if no current subscriptions. this helps prevent
+      ; the subscribed map from (for example) endlessly filling up with massive
+      ; amounts of entries with no subscriptions. this could easily happen over time
+      ; naturally for applications with long uptimes.
+      (update :subscribed
+              (fn [subscribed]
+                (if (empty? (get subscribed subscriber-key))
+                  (dissoc subscribed subscriber-key)
+                  subscribed)))))
+
+(defn- clean-up-unneeded-hashes
+  "Remove hashes for view-sigs which do not have any unsubscribers."
+  [view-system view-sig]
+  (if-not (get (:subscribers view-system) view-sig)
+    (update view-system :hashes dissoc view-sig)
+    view-system))
 
 (defn unsubscribe!
-  [view-system namespace view-id parameters subscriber-key]
+  "Removes a subscription to a view identified by view-sig for a subscriber
+   identified by subscriber-key. Additional context info can be passed in,
+   which will be passed to the view-system's namespace-fn (if provided)."
+  [view-system {:keys [view-id parameters] :as view-sig} subscriber-key context]
+  (trace "unsubscribing from view" view-sig subscriber-key)
   (swap! view-system
          (fn [vs]
-           (-> vs
-               (update-in [:subscribed subscriber-key] disj [namespace view-id parameters])
-               (remove-from-subscribers [namespace view-id parameters] subscriber-key)
-               (update-hash! [namespace view-id parameters] nil))))) ;; see note #2 in NOTES.md
+           (let [namespace (view-namespace vs view-sig subscriber-key context)
+                 view-sig  (->view-sig namespace view-id parameters)]
+             (-> vs
+                 (remove-from-subscribed view-sig subscriber-key)
+                 (remove-from-subscribers view-sig subscriber-key)
+                 (assoc-hash view-sig nil) ;; see note #2 in NOTES.md
+                 (clean-up-unneeded-hashes view-sig)))))
```

Comment:
Yes, doubt it's going to be a big issue but I don't see a problem with leaving it.